crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -791,6 +791,11 @@ impl<'db> TypeInferenceBuilder<'db> {
         let mut bound_ty = ty;
         let declared_ty = declarations_ty(self.db, declarations, undeclared_ty).unwrap_or_else(
             |(ty, conflicting)| {
+                if let Some(undeclared_ty) = undeclared_ty {
+                    if conflicting.iter().any(|ty| ty == &undeclared_ty) {
+                        return ty;
+                    }
+                }
```

Comment:
I think this has an issue: there could be three or more conflicting types, e.g. `int`, `str`, and `Unknown`; this should still emit a diagnostic for the conflict between `int` and `str`.

It's also a little weird that we just search for `Unknown`, because there could be an actual annotation `x: Foo` in one path, where `Foo` resolves to `Unknown` (e.g. because it is imported from a module we can't find), and we would still treat this `Unknown` the same as the undeclared-path `Unknown`.

I think it would be easier to make a more correct fix if we do it inside `declarations_ty` instead, so that in the cases where we don't want to emit the diagnostic, we don't even return an `Err` result at all.

Reply:
Ah yes, thanks for pointing that out. I've pushed a commit that moves the logic to `declarations_ty` which doesn't consider the `undeclared_ty` when looking for conflicting types. It _does_ add the type to the final declared type.